libbeat/beat/instrumentation.go

```diff
@@ -0,0 +1,241 @@
+	return p != nil && p.Enabled
+}
+
+// IsEnabled indicates whether heap profiling is enabled
+func (p *HeapProfiling) IsEnabled() bool {
+	return p != nil && p.Enabled
+}
+
+// HeapProfiling holds config information about heap profiling of the APM Server
+type HeapProfiling struct {
+	Enabled  bool          `config:"enabled"`
+	Interval time.Duration `config:"interval" validate:"positive"`
+}
+
+// CreateInstrumentation configures and returns an instrumentation object for tracing
+func CreateInstrumentation(cfg *common.Config, info Info) (*Instrumentation, error) {
+	if !cfg.HasField("instrumentation") {
+		return nil, nil
+	}
+
+	instrumentation, err := cfg.Child("instrumentation", -1)
+	if err != nil {
+		return nil, nil
+	}
+
+	config := InstrumentationConfig{}
+
+	if instrumentation == nil {
+		instrumentation = common.NewConfig()
+	}
+	err = instrumentation.Unpack(&config)
+
+	if err != nil {
+		return nil, fmt.Errorf("could not create tracer, err: %v", err)
+	}
+
+	return initTracer(config, info)
+}
+
+func initTracer(cfg InstrumentationConfig, info Info) (*Instrumentation, error) {
+
+	logger := logp.NewLogger("tracing")
+
+	if !cfg.IsEnabled() {
+		os.Setenv("ELASTIC_APM_ACTIVE", "false")
+		logger.Infof("APM instrumentation is disabled")
+		return nil, nil
+	} else {
+		os.Setenv("ELASTIC_APM_ACTIVE", "true")
+		logger.Infof("APM instrumentation is enabled")
+	}
+
+	if cfg.Profiling.CPU.IsEnabled() {
+		interval := cfg.Profiling.CPU.Interval
+		duration := cfg.Profiling.CPU.Duration
+		logger.Infof("CPU profiling: every %s for %s", interval, duration)
+		os.Setenv("ELASTIC_APM_CPU_PROFILE_INTERVAL", fmt.Sprintf("%dms", int(interval.Seconds()*1000)))
+		os.Setenv("ELASTIC_APM_CPU_PROFILE_DURATION", fmt.Sprintf("%dms", int(duration.Seconds()*1000)))
+	}
+	if cfg.Profiling.Heap.IsEnabled() {
+		interval := cfg.Profiling.Heap.Interval
+		logger.Infof("Heap profiling: every %s", interval)
+		os.Setenv("ELASTIC_APM_HEAP_PROFILE_INTERVAL", fmt.Sprintf("%dms", int(interval.Seconds()*1000)))
+	}
+
+	var tracerTransport transport.Transport
+	var tracerListener net.Listener
+
+	if cfg.Hosts == nil && info.Name == "apm-server" {
+		pipeListener := pipe.NewListener()
+		pipeTransport, err := transport.NewHTTPTransport()
+		if err != nil {
+			return nil, err
+		}
+		pipeTransport.SetServerURL(&url.URL{Scheme: "http", Host: "localhost:8200"})
+		pipeTransport.Client.Transport = &http.Transport{
+			DialContext:     pipeListener.DialContext,
+			MaxIdleConns:    100,
+			IdleConnTimeout: 90 * time.Second,
+		}
+		tracerTransport = pipeTransport
+		// the traceListener will allow APM Server to create an ad-hoc server for tracing
+		tracerListener = pipeListener
+
+	} else {
+		t, err := transport.NewHTTPTransport()
+		if err != nil {
+			return nil, err
+		}
+		hosts := cfg.Hosts
+		if hosts == nil {
+			hosts = []*url.URL{{
+				Scheme: "http",
+				Host:   "localhost:8200",
+			}}
+		}
+		t.SetServerURL(hosts...)
```

Comment:
```suggestion
		if len(cfg.Hosts) > 0 {
			t.SetServerURL(cfg.Hosts...)
		}
```

NewHTTPTransport initialises the URLs with localhost:8200.